README.md
# Randomly-Generate-Test-Cases-for-2D-Convolution

Simple as it is, I implement it in three distinct languages: Kotlin, Python (by PyTorch), and Haskell, which brings me some interesting insights.

## Dependencies for Haskell

```
dependencies:
- base >= 4.7 && < 5
- mtl
- containers
- filepath
- directory
- time
- string-interpolate
- array
```

## Dependencies for Python

```
pytorch
```


Gen-Conv-Cases-Haskell/CHANGELOG.md
# Changelog for `Gen-Conv-Cases-Haskell`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


ConvTestCases.kt
import java.io.File
import kotlin.math.pow
import kotlin.random.Random

// randomly generate a matrix of shape [height, width]
fun randMatrix(height: Int, width: Int): List<List<Float>> {
    assert(height > 0)
    assert(width > 0)

    // expRange is exclusive
    fun randFloatValidBits(expRange: Int): Float {
        val floatExpBits = 8
        assert(expRange in 1..(1 shl floatExpBits))
        val f = Random.nextFloat() + 1 // \in [1, 2)
        val sign = if (Random.nextBoolean()) -1 else 1
        val exp = Random.nextInt(-expRange + 1, expRange)
        return sign * f * 2.0f.pow(exp)
    }

    return List(height) {
        List(width) {
            randFloatValidBits(expRange = 50) // ensures no 'inf' will be produced
        }
    }
}

// each sublist of matrix/kernel must have the same size, otherwise the assertions will fail
fun convolution(matrix: List<List<Float>>, kernel: List<List<Float>>): List<List<Float>> {
    val matrixHeight = matrix.size
    val matrixWidth = matrix.first().size
    val kernelHeight = kernel.size
    val kernelWidth = kernel.first().size
    assert(kernelHeight > 0)
    assert(kernelWidth > 0)
    assert(matrixHeight >= kernelHeight)
    assert(matrixWidth >= kernelWidth)
    assert(matrix.all { it.size == matrixWidth })
    assert(kernel.all { it.size == kernelWidth })
    return List(matrixHeight - kernelHeight + 1) { h ->
        List(matrixWidth - kernelWidth + 1) { w ->
            (0 until kernelHeight).map { i ->
                (0 until kernelWidth).map { j ->
                    matrix[i + h][j + w] * kernel[i][j]
                }.sum()
            }.sum()
        }
    }
}

// output: (matrix, kernel, convolution_output)
fun generateOneTestCase(
    matrixHeight: Int, matrixWidth: Int, kernelHeight: Int, kernelWidth: Int
): Triple<List<List<Float>>, List<List<Float>>, List<List<Float>>> {
    assert(matrixHeight > 0)
    assert(matrixWidth > 0)
    assert(kernelHeight > 0)
    assert(kernelWidth > 0)
    val matrix = randMatrix(matrixHeight, matrixWidth)
    val kernel = randMatrix(kernelHeight, kernelWidth)
    val output = convolution(matrix, kernel)
    return Triple(matrix, kernel, output)
}

fun generateTestCases(
    matrixShapeList: List<Pair<Int, Int>>,
    kernelShapeList: List<Pair<Int, Int>>
): List<Triple<List<List<Float>>, List<List<Float>>, List<List<Float>>>> =
    matrixShapeList.flatMap { (mh, mw) ->
        kernelShapeList.filter { (kh, kw) -> mh >= kh && mw >= kw }
            .map { (kh, kw) ->
                generateOneTestCase(mh, mw, kh, kw)
            }
    }

fun printMatrix(matrix: List<List<Float>>, path: String) =
    File(path).printWriter().use { f ->
        matrix.forEach { row ->
            f.println(row.map { it.toString() }.reduce { a, b -> "$a $b" })
        }
    }

fun printOneTestCase(matrices: Triple<List<List<Float>>, List<List<Float>>, List<List<Float>>>, dirPath: String) {
    val (matrix, kernel, output) = matrices
    val matrixHeight = matrix.size
    val matrixWidth = matrix.first().size
    val kernelHeight = kernel.size
    val kernelWidth = kernel.first().size
    val subDirName = "${matrixHeight}x${matrixWidth}_${kernelHeight}x${kernelWidth}"
    val subDirPath = "$dirPath/$subDirName"
    File(subDirPath).mkdir()
    printMatrix(matrix, "$subDirPath/matrix.txt")
    printMatrix(kernel, "$subDirPath/kernel.txt")
    printMatrix(output, "$subDirPath/output.txt")
    println("$subDirName Finished")
}

fun printTestCases(tripleList: List<Triple<List<List<Float>>, List<List<Float>>, List<List<Float>>>>, dirPath: String) =
    tripleList.forEach { printOneTestCase(it, dirPath) }

fun main() {
    val beginTime = System.currentTimeMillis()
    val testCasesPath = "test_cases"
    File(testCasesPath).mkdir()
    val tripleList = generateTestCases(
        listOf(
            Pair(32, 32), Pair(64, 64), Pair(128, 128), Pair(256, 256), Pair(512, 512), Pair(1024, 1024)
        ),
        listOf(
            Pair(16, 16), Pair(64, 64)
        )
    )
    printTestCases(tripleList, testCasesPath)
    val endTime = System.currentTimeMillis()
    println("total time: ${endTime - beginTime}ms")
}


conv_test_cases.py
from typing import List, Tuple
import torch
import torch.nn.functional as F
import os
import shutil
import time

CASES_PER_SETTING = 1
INPUT_SHAPE = [(512, 1024), (1024, 1024), (1024, 2048), (2048, 2048)]
WEIGHT_SHAPE = [(16, 16), (32, 32), (64, 64)]
OUTPUT_DIR = 'conv_test_cases'


# matrix is a 2d matrix
def write_matrix(matrix: torch.FloatTensor, path: str):
    with open(path, 'w') as file:
        height = matrix.size(0)
        width = matrix.size(1)
        file.write(str(height)+' '+str(width)+'\n')
        for h in range(height):
            for w in range(width):
                file.write(str(matrix[h][w].item())+' ')
            file.write('\n')


def generate_cases(input_shape: List[Tuple[int]], weight_shape: List[Tuple[int]]):
    for i in input_shape:
        for w in weight_shape:
            for c in range(CASES_PER_SETTING):
                if i[0] < w[0] or i[1] < w[1]:
                    continue
                begin_time = time.time()
                input = torch.rand(1, 1, i[0], i[1])
                weight = torch.rand(1, 1, w[0], w[1])
                output = F.conv2d(input, weight, padding='valid')
                sub_dir_path = os.path.join(OUTPUT_DIR,
                                            str(i[0])+'x'+str(i[1])+'_' +
                                            str(w[0])+'x'+str(w[1])+'_'+str(c+1))
                os.mkdir(sub_dir_path)
                write_matrix(input[0][0], os.path.join(
                    sub_dir_path, 'input.txt'))
                write_matrix(weight[0][0], os.path.join(
                    sub_dir_path, 'weight.txt'))
                write_matrix(output[0][0], os.path.join(
                    sub_dir_path, 'output.txt'))
                end_time = time.time()
                print('input shape '+str(i)+' weight shape '+str(w)+': case '+str(c+1) +
                      '/'+str(CASES_PER_SETTING)+' finished')
                print('time elapsed: '+str(end_time-begin_time)+'s')


def main():
    begin_time = time.time()
    torch.manual_seed(19260817)
    if os.path.exists(OUTPUT_DIR) and os.path.isdir(OUTPUT_DIR):
        shutil.rmtree(OUTPUT_DIR)
    os.mkdir(OUTPUT_DIR)
    generate_cases(INPUT_SHAPE, WEIGHT_SHAPE)
    end_time = time.time()
    print('total time: '+str(end_time-begin_time)+'s')


if __name__ == '__main__':
    main()


Gen-Conv-Cases-Haskell/app/Main.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE ImpredicativeTypes #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE QuasiQuotes #-}

module Main (main) where

import Control.Concurrent (forkFinally, putMVar, takeMVar)
import Control.Concurrent.MVar (newEmptyMVar)
import Control.Monad (forM_, replicateM_)
import Control.Monad.State.Strict (State, runState, state)
import Data.Array.MArray (Ix, MArray (newArray), readArray, writeArray)
import Data.Array.ST (runSTUArray)
import Data.Array.Unboxed (IArray, UArray, (!))
import Data.Functor ((<&>))
import Data.Int (Int64)
import Data.String.Interpolate (i)
import Data.Time.Clock (diffUTCTime, getCurrentTime)
import System.Directory (createDirectoryIfMissing)
import System.Directory.Internal.Prelude (getArgs)
import System.FilePath ((<.>), (</>))

type RandomNumberType = Float

rng :: State Int64 Int64
rng =
  state
    ( \s ->
        let !next = s * 1103515245 + 12345
         in ((next `div` 65536) `mod` 32768, next)
    )

randMatrix :: (MArray a e m, Ix t, Num t, Num e) => t -> t -> Int64 -> m (a t e)
randMatrix !height !width !seed = do
  arr <- newArray (0, height * width - 1) 0
  let fill !ix !s =
        ( if ix >= height * width
            then return arr
            else
              let castInt int = fromIntegral (int `div` 128) - 128
                  !(float, nextSeed) = runState (rng <&> castInt) s
               in ( do
                      writeArray arr ix float
                      fill (ix + 1) nextSeed
                  )
        )
  fill 0 seed

convolution ::
  ( MArray a1 t m,
    Ix i,
    Num i,
    Num t,
    Enum i,
    IArray a2 t,
    IArray a3 t
  ) =>
  a2 i t ->
  (i, i) ->
  a3 i t ->
  (i, i) ->
  m (a1 i t)
convolution !input (!inputHeight, !inputWidth) !weight (!weightHeight, !weightWidth) = do
  let !(outputHeight, outputWidth) =
        (inputHeight - weightHeight + 1, inputWidth - weightWidth + 1)
  !output <- newArray (0, outputHeight * outputWidth - 1) 0
  forM_
    [0 .. inputHeight - weightHeight]
    ( \oh ->
        forM_
          [0 .. inputWidth - weightWidth]
          ( \ow ->
              forM_
                [0 .. weightHeight - 1]
                ( \wh ->
                    forM_
                      [0 .. weightWidth - 1]
                      ( \ww -> do
                          let !(ih, iw) = (oh + wh, ow + ww)
                          let !srcI = input ! (ih * inputWidth + iw)
                          let !srcW = weight ! (wh * weightWidth + ww)
                          modifyArray output (oh * outputWidth + ow) (+ (srcI * srcW))
                      )
                )
          )
    )
  return $! output
  where
    modifyArray !arr !ix !f = do
      !origin <- readArray arr ix
      writeArray arr ix $! f origin

genOneCase ::
  (Ix a, Num a, Enum a) =>
  (a, a, a, a) ->
  (Int64, Int64) ->
  ( UArray a RandomNumberType,
    UArray a RandomNumberType,
    UArray a RandomNumberType
  )
genOneCase
  ( !inputHeight,
    !inputWidth,
    !weightHeight,
    !weightWidth
    )
  (!seed1, !seed2) =
    let !input = runSTUArray $! randMatrix inputHeight inputWidth seed1
        !weight = runSTUArray $! randMatrix weightHeight weightWidth seed2
        !output =
          runSTUArray $!
            convolution
              input
              (inputHeight, inputWidth)
              weight
              (weightHeight, weightWidth)
     in (input, weight, output)

outputMatrix :: UArray Int RandomNumberType -> (Int, Int) -> FilePath -> String -> IO ()
outputMatrix !matrix (!height, !width) !dir !name = do
  let !path = dir </> name <.> "txt"
  writeFile path [i|#{height} #{width}\n|]
  appendFile path $! show matrix

outputOneCase :: (Int, Int, Int, Int) -> FilePath -> (Int64, Int64) -> IO ()
outputOneCase
  ( !inputHeight,
    !inputWidth,
    !weightHeight,
    !weightWidth
    )
  !path
  (!seed1, !seed2) = do
    let !(input, weight, output) =
          genOneCase (inputHeight, inputWidth, weightHeight, weightWidth) (seed1, seed2)

    let !outputDirPath =
          path
            </> [i|#{inputHeight}x#{inputWidth}_#{weightHeight}x#{weightWidth}|]

    createDirectoryIfMissing True outputDirPath

    let !(outputHeight, outputWidth) =
          (inputHeight - weightHeight + 1, inputWidth - weightWidth + 1)

    outputMatrix input (inputHeight, inputWidth) outputDirPath "input"
    outputMatrix weight (weightHeight, weightWidth) outputDirPath "weight"
    outputMatrix output (outputHeight, outputWidth) outputDirPath "output"

outputCases :: [(Int, Int, Int, Int)] -> FilePath -> IO ()
outputCases !shapeList !path = do
  !wg <- newEmptyMVar
  forM_
    (zip [1 ..] shapeList)
    ( \(!ix, !shape) ->
        forkFinally
          (outputOneCase shape path (ix * 2, ix * 2 + 1))
          ( \case
              Right _ -> putMVar wg ()
              Left _ -> error "an error occured"
          )
    )
  replicateM_ (length shapeList) $! takeMVar wg

main :: IO ()
main = do
  start <- getCurrentTime
  let shapeList =
        [ (512, 512, 8, 8),
          (512, 512, 16, 16),
          (512, 512, 32, 32),
          (512, 512, 64, 64)
        ]
  outputPath : _ <- getArgs
  outputCases shapeList outputPath
  end <- getCurrentTime
  putStrLn [i|#{end `diffUTCTime` start} elapsed.|]